luigi/worker.py

```diff
@@ -775,7 +775,7 @@ def _add(self, task, is_complete):
             runnable=runnable,
             priority=task.priority,
             resources=task.process_resources(),
-            params=task.to_str_params(),
+            params=task.to_str_params(only_visible=True),
```

Comment:
I'm not sure I understand here why you set the default to `false`, but force the worker to pass `true`. I feel that whatever the result may be, it shouldn't affect people who don't have this value set on their Parameters.

Thus if visible is set to true, then `only_visible` should be `true` by default.

Reply:
Thank you for your response!

> I feel that whatever the result may be, it shouldn't affect people who don't have this value set on their Parameters.

It will not affect such people because in luigi.Parameter's constructor visible value by default is True

> I'm not sure I understand here why you set the default to false, but force the worker to pass true

I set the default value to false, because method `to_str_params` is called from other places and in such cases it expects to recieve all parameters from task regardless of values like significant or visible.

For example, 
```
    def _announce_scheduling_failure(self, task, expl):
            self._scheduler.announce_scheduling_failure(
                worker=self._id,
                task_name=str(task),
                family=task.task_family,
                params=task.to_str_params(only_significant=True), # regadless of visible, will return only significant params
                expl=expl,
                owners=task._owner_list(),
            )
```

```
     self._scheduler.add_task(
                worker=self._id,
                task_id=task.task_id,
                module=get_work_response.get('task_module'),
                family=get_work_response['task_family'],
                params=task.to_str_params(), # as before return all params 
                status=RUNNING,
                batch_id=get_work_response['batch_id'],
            )
```
>Thus if visible is set to true, then only_visible should be true by default.
In general, i tried to copy `significant` logic but for another purpose and `visible` is similar to it

```
def __init__(self, default=_no_value, is_global=False, significant=True, description=None,
                 config_path=None, positional=True, always_in_help=False, batch_method=None):
```
```
    def __init__(self, default=_no_value, is_global=False, significant=True, description=None,
                 config_path=None, positional=True, always_in_help=False, batch_method=None, visible=True):
```

Reply:
@dlstadther Thank you for your response!

> I feel that whatever the result may be, it shouldn't affect people who don't have this value set on their Parameters.

It will not affect such people because in luigi.Parameter's constructor visible value by default is True

> I'm not sure I understand here why you set the default to false, but force the worker to pass true

I set the default value to false, because method `to_str_params` is called from other places and in such cases it expects to recieve all parameters from task regardless of values like significant or visible.

For example, 
```
    def _announce_scheduling_failure(self, task, expl):
            self._scheduler.announce_scheduling_failure(
                worker=self._id,
                task_name=str(task),
                family=task.task_family,
                params=task.to_str_params(only_significant=True), # regadless of visible, will return only significant params
                expl=expl,
                owners=task._owner_list(),
            )
```

```
     self._scheduler.add_task(
                worker=self._id,
                task_id=task.task_id,
                module=get_work_response.get('task_module'),
                family=get_work_response['task_family'],
                params=task.to_str_params(), # as before return all params 
                status=RUNNING,
                batch_id=get_work_response['batch_id'],
            )
```
>Thus if visible is set to true, then only_visible should be true by default.

In general, i tried to copy `significant` logic but for another purpose and `visible` is similar to it

```
def __init__(self, default=_no_value, is_global=False, significant=True, description=None,
                 config_path=None, positional=True, always_in_help=False, batch_method=None):
```
```
    def __init__(self, default=_no_value, is_global=False, significant=True, description=None,
                 config_path=None, positional=True, always_in_help=False, batch_method=None, visible=True):
```

Reply:
So then, if i understand this correctly... this `_add()` method is what the worker uses to push the definition of each task to the web view? So by removing non-visible params from the task's tostr, you remove them from visibility?

Are there any negative side-effects if i have two nearly identical tasks which only differ in username/password? They would have the same result from `to_str_params()` when they are not the same task.

Reply:
>So then, if i understand this correctly... this _add() method is what the worker uses to push the definition of each task to the web view? So by removing non-visible params from the task's tostr, you remove them from visibility?

Exactly. In my case, the main purpose was to hide password, username, host and port only from web view tasks like that:

```
class SomeAction(luigi.Task):
    password = luigi.Parameter(visible=False)
    username= luigi.Parameter(visible=False)
    host= luigi.Parameter(visible=False)
    port= luigi.Parameter(visible=False)
    id = luigi.IntParameter()
    date = luigi.DateParameter()

    def output(self):
    return luigi.PostgresTarget(...)
...
```
In the result, in web page visible only ID and Date. I tested it on our dev place with different parameters and it works fine for me.
Also, for additional testing i pass to luigi.interface.run my own worker_scheduler_factory to produce custom scheduler\workers and i place some logs into it to control task params 

>Are there any negative side-effects if i have two nearly identical tasks which only differ in username/password? They would have the same result from to_str_params() when they are not the same task.

Probably, i don't correctly understand what do you mean by ` They would have the same result from to_str_params() when they are not the same task`. For such tasks, to_str_params would have some affect only for params where visible=False and only for web page, not for the task instance

Reply:
Gotcha. I think i understand.